addons/l10n_se_ocr/models/account_move.py

```diff
@@ -0,0 +1,85 @@
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _get_invoice_reference_se_ocr2(self, number):
+        self.ensure_one()
+
+        reference = self.env['l10n_se.luhn'].append(str(number))
+
+        return reference
+
+    def _get_invoice_reference_se_ocr3(self, number):
+        self.ensure_one()
+
+        reference = self.env['l10n_se.luhn'].append(str(number) + str(len(str(number))+2)[:1])
+
+        return reference
+
+    def _get_invoice_reference_se_ocr4(self, number):
+        self.ensure_one()
+
+        ocr_length = self.journal_id.invoice_ocr_length
+
+        if ocr_length < 5:
+            raise UserError(_("OCR Reference Number length need to be greater then 5. Please correct settings under invoice journal settings."))
+
+        if len(str(number))+1 > ocr_length:
+            raise UserError(_("OCR Reference Number length is greater than allowed. Allowed length in invoice journal setting is %s.") % str(ocr_length))
+
+        reference = self.env['l10n_se.luhn'].append(str(number).ljust(ocr_length-1, '0'))
+
+        return reference
+
+    def _get_invoice_reference_se_ocr2_invoice(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr2(str(self.id))
+
+    def _get_invoice_reference_se_ocr3_invoice(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr3(str(self.id))
+
+    def _get_invoice_reference_se_ocr4_invoice(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr4(str(self.id))
+
+    def _get_invoice_reference_se_ocr2_partner(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr2(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    def _get_invoice_reference_se_ocr3_partner(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr3(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    def _get_invoice_reference_se_ocr4_partner(self):
+        self.ensure_one()
+
+        return self._get_invoice_reference_se_ocr4(self.partner_id.ref if str(self.partner_id.ref).isdecimal() else str(self.partner_id.id))
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        if hasattr(super(AccountMove, self), "_onchange_partner_id"):
+            super(AccountMove, self)._onchange_partner_id()
+
+        """ If Vendor Bill and Vendor OCR is set, add it. """
+        if self.partner_id and self.type == 'in_invoice' and self.partner_id.use_vendor_ocr and self.partner_id.default_vendor_ocr:
+            self.invoice_payment_ref = self.partner_id.default_vendor_ocr
+
+    @api.onchange('invoice_payment_ref')
+    def _onchange_invoice_payment_ref(self):
+        if hasattr(super(AccountMove, self), "_onchange_payment_ref"):
```

Comment:
Well AccountMove doesn't have it. So if I don't check for it, it will cast an exception. There is two ways of doing it.

```
try: 
    super(AccountMove, self)._onchange_invoice_payment_ref()
catch:
# Do nothing...
```

or you could do it with hasattr function used. The choice is yours since I'm not sure what you think is better. Just decide what method you prefer and I will change it quickly.

Reply:
https://github.com/odoo/odoo/blob/f80613fdbcc7aba8d70894defd7bce6ecc5a42f6/addons/account/models/account_move.py#L391

Reply:
My fault.. I need to check this since it throws an exception if removed. However I need to check why. I take a look at it later today.

Reply:
fixed